### DISCUSSION TITLE
Bidirectional many-to-many relationship Users <-> Roles <-> Permissions on SEPARATE FILES

### DISCUSSION THREAD

**Discussion by ricardo-reis-1970** [pydantic/pydantic#3015] [Question]
# Environment
- Python: 3.9.0
- Pydantic: 1.8.2

# The issue
I keep seeing the same examples over and over again here, in documentation, in short answers, etc. These always fail to match reality. We are told that each entity should be in its own separate file, but bidirectional many-to-many relationships fail to work.

I (wanted to) have the following files under my folder `schemas`:
## `permission.py`
```python
from typing import List
from . import BaseModel, CreatedModel
from .role import RoleSchema

class PermissionBase(BaseModel):
    name: str
    description: str

class PermissionCreate(PermissionBase):
    pass

class PermissionSchema(PermissionBase, CreatedModel):
    """Schema validation for Permission"""
    roles: List[RoleSchema] = []

    class Config:
        orm_mode = True
```

## `user.py`
```python
from typing import Optional, List
from pydantic import SecretStr
from . import BaseModel, CreatedModel
from .role import RoleSchema


class UserBase(BaseModel):
    username: str
    gecos: Optional[str] = None
    email: Optional[str] = None
    is_active: Optional[bool] = False
    is_authenticated: Optional[bool] = False

class UserCreate(UserBase):
    password: str

class UserSchema(UserBase, CreatedModel):
    """Schema validation for User"""
    roles: List[RoleSchema] = []

    class Config:
        orm_mode = True
```

## `role.py`
```python
from typing import Optional, List, ForwardRef
from . import BaseModel, CreatedModel

class RoleBase(BaseModel):
    name: str
    description: str

class RoleCreate(RoleBase):
    pass

class RoleSchema(RoleBase, CreatedModel):
    """Schema validation for Role"""
    permissions: List[ForwardRef('PermissionSchema')] = []
    users: List[ForwardRef('UserSchema')] = []

    class Config:
        orm_mode = True
```

Please notice that I import `RoleSchema` in both `user.py` and `permission.py`. Due to this, I can't import any of these two in `role.py`, otherwise I'd be falling in a circular import. The solution — I thought — was to make both `permissions` and `users` to be defined as `ForwardRef`.

This is all good, except that at somepoint I must do:
```python
RoleSchema.update_forward_refs()
```
and this, wherever I try and do it, will always produce the error:
```
  [...]
  File "pydantic\main.py", line 832, in pydantic.main.BaseModel.update_forward_refs
  File "pydantic\typing.py", line 382, in pydantic.typing.update_field_forward_refs

  File "pydantic\typing.py", line 70, in pydantic.typing.evaluate_forwardref
    'Coroutine',
  File "C:\Program Files\Python39\lib\typing.py", line 493, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
NameError: name 'PermissionSchema' is not defined
```

I tried doing the `update_forward_refs` in the `role.py` file, I tried in the schemas' `__init__.py` (after importing all the entities) and I even tried this in the `main.py` file. The error always happens.

I thought of just leaving the forward references unresolved, but obviously this is coming to bite my arse later. For now, the only way to have this working seems to be this:

## `role.py` (workaround)
```python
from typing import Optional, List, ForwardRef
from . import BaseModel, CreatedModel

class RoleBase(BaseModel):
    name: str
    description: str

class RoleCreate(RoleBase):
    pass

class RoleSchema(RoleBase, CreatedModel):
    """Schema validation for Role"""
    permissions: List[ForwardRef('PermissionSchema')] = []
    users: List[ForwardRef('UserSchema')] = []

    class Config:
        orm_mode = True

class PermissionBase(BaseModel):
    name: str
    description: str

class PermissionCreate(PermissionBase):
    pass

class PermissionSchema(PermissionBase, CreatedModel):
    """Schema validation for Permission"""
    roles: List[RoleSchema] = []

    class Config:
        orm_mode = True

class UserBase(BaseModel):
    username: str
    gecos: Optional[str] = None
    email: Optional[str] = None
    is_active: Optional[bool] = False
    is_authenticated: Optional[bool] = False

class UserCreate(UserBase):
    password: str

class UserSchema(UserBase, CreatedModel):
    """Schema validation for User"""
    roles: List[RoleSchema] = []

    class Config:
        orm_mode = True

RoleSchema.update_forward_refs()
```

This means including the three entities in the same file. Either I have one single file for all entities, or I have users and permissions duplicated in `role.py`.

This does not make much sense to me. I was under the impression that the whole point of `ForwardRef` and referring to entities by string name was precisely this kind of situation.

# Finally
I used to have this in Flask, Flask-SQLAlchemy and marshmallow. It all worked. I'm moving into FastAPI, SLQAlchemy and Pydantic.

What am I doing wrong?

**Comment by cblack34** [comment]
I am also looking for the best solution to this problem.

I'm sure there is a better way but the best I've found so far is here: https://stackoverflow.com/a/63425646

Basically import at bottom and then do .update_forward_refs()

Hope that helps and hope someone suggests a better answer.
